internal/datastore/proxy/schemacaching/intervaltracker.go

```diff
@@ -0,0 +1,162 @@
+package schemacaching
+
+import (
+	"sync"
+	"time"
+
+	"golang.org/x/exp/slices"
+
+	"github.com/authzed/spicedb/pkg/datastore"
+)
+
+// intervalTracker is a specialized type for tracking a value over a set of
+// revision-based time intervals.
+type intervalTracker[T any] struct {
+	// sortedEntries are the entries in the interval tracker, sorted with the *latest* being *first*.
+	sortedEntries []intervalTrackerEntry[T]
+
+	// lock is the sync lock for the tracker.
+	lock sync.RWMutex
+}
+
+// intervalTrackerEntry is a single entry in the interval tracker, over a
+// period of revisions.
+type intervalTrackerEntry[T any] struct {
+	// created is the point at which the entry was created.
+	created time.Time
+
+	// value is the value for the tracked interval.
+	value T
+
+	// startingRevision at which the interval begins.
+	startingRevision datastore.Revision
+
+	// endingRevision is the *exclusive* revision at which the interval ends. If not specified,
+	// then the interval is open until the last checkpoint revision.
+	endingRevisionOrNil datastore.Revision
+}
+
+// newIntervalTracker creates a new interval tracker.
+func newIntervalTracker[T any]() *intervalTracker[T] {
+	return &intervalTracker[T]{
+		sortedEntries: make([]intervalTrackerEntry[T], 0, 1),
+	}
+}
+
+// removeStaleIntervals removes all fully-defined intervals that were created at least window-time ago.
+// Returns true if *all* intervals were removed, and the tracker is now empty.
+func (it *intervalTracker[T]) removeStaleIntervals(window time.Duration) bool {
+	threshold := time.Now().Add(-window)
+
+	it.lock.Lock()
+	defer it.lock.Unlock()
+
+	it.sortedEntries = slices.DeleteFunc(it.sortedEntries, func(entry intervalTrackerEntry[T]) bool {
+		// The open-ended entry always remains.
+		if entry.endingRevisionOrNil == nil {
+			return false
+		}
+
+		return entry.created.Before(threshold)
+	})
+
+	return len(it.sortedEntries) == 0
+}
+
+// lookup performs lookup of the value in the tracker at the specified revision. lastCheckpointRevision is the
+// bound to use for the ending revision for the unbounded entry in the tracker (if any).
+func (it *intervalTracker[T]) lookup(revision datastore.Revision, lastCheckpointRevision datastore.Revision) (T, bool) {
+	it.lock.RLock()
+	defer it.lock.RUnlock()
+
+	// NOTE: The sortedEntries slice is sorted from latest to least recent, which is opposite that expected
+	// by BinarySearchFunc, so all the returned values below are "inverted".
```

Comment:
nit: BinarySearchFunc doesn't have an opinion on slice order aside from what the cmp function says

Reply:
But it does have an opinion based on the cmp function itself, which is why I mention its inverted

Reply:
> The slice must be sorted in increasing order, where "increasing" is defined by cmp. cmp should return 0 if the slice element matches the target, a negative number if the slice element precedes the target, or a positive number if the slice element follows the target. cmp must implement the same ordering as the slice, such that if cmp(a, t) < 0 and cmp(b, t) >= 0, then a must precede b in the slice.